[tune] raises RuntimeError about non-training futures

<!--Please include [tune], [rllib], [autoscaler] etc. in the issue title if relevant-->

### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):* Ray 1.3.0 Python 3.8 Ubuntu Server 18.04 PyTorch 1.7.1

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

I tried to run the script named `mnist.py`:

```python
import argparse
import os
import torch
import torch.optim as optim

import ray
from ray import tune
from ray.tune.schedulers import ASHAScheduler
from ray.tune.examples.mnist_pytorch import (train, test, get_data_loaders,
                                             ConvNet)

# Change these values if you want the training to run quicker or slower.
EPOCH_SIZE = 512
TEST_SIZE = 256

# Training settings
parser = argparse.ArgumentParser(description="PyTorch MNIST Example")
parser.add_argument("--use-gpu", action="store_true", default=True, help="enables CUDA training")
parser.add_argument("--ray-address", type=str, help="The Redis address of the cluster.")
parser.add_argument("--smoke-test", action="store_true", help="Finish quickly for testing")

# Below comments are for documentation purposes only.
# yapf: disable
# __trainable_example_begin__
class TrainMNIST(tune.Trainable):
  def setup(self, config):
    use_cuda = config.get("use_gpu") and torch.cuda.is_available()
    self.device = torch.device("cuda" if use_cuda else "cpu")
    self.train_loader, self.test_loader = get_data_loaders()
    self.model = ConvNet().to(self.device)
    self.optimizer = optim.SGD(
        self.model.parameters(),
        lr=config.get("lr", 0.01),
        momentum=config.get("momentum", 0.9))

  def step(self):
    train(self.model, self.optimizer, self.train_loader, device=self.device)
    acc = test(self.model, self.test_loader, self.device)
    return {"mean_accuracy": acc}

  def save_checkpoint(self, checkpoint_dir):
    checkpoint_path = os.path.join(checkpoint_dir, "model.pth")
    torch.save(self.model.state_dict(), checkpoint_path)
    return checkpoint_path

  def load_checkpoint(self, checkpoint_path):
    self.model.load_state_dict(torch.load(checkpoint_path))

  # this is currently needed to handle Cori GPU multiple interfaces
  def current_ip(self):
    import socket
    hostname = socket.getfqdn(socket.gethostname())
    self._local_ip = socket.gethostbyname(hostname)
    return self._local_ip 


if __name__ == "__main__":
  # ip_head and redis_passwords are set by ray cluster shell scripts
  print(os.environ["ip_head"], os.environ["redis_password"])
  ray.init(address=os.environ["ip_head"], _redis_password=os.environ["redis_password"])
  sched = ASHAScheduler(metric="mean_accuracy", mode="max")
  analysis = tune.run(TrainMNIST,
                      scheduler=sched,
                      stop={"mean_accuracy": 0.99,
                            "training_iteration": 100},
                      resources_per_trial={"cpu":10, "gpu": 1},
                      num_samples=128,
                      checkpoint_at_end=True,
                      config={"lr": tune.uniform(0.001, 1.0),
                              "momentum": tune.uniform(0.1, 0.9),
                             "use_gpu": True})
  print("Best config is:", analysis.get_best_config(metric="mean_accuracy", mode="max"))
```

Then I started ray on two GPU servers by the following commands:

```bash
# On the first server
ray start --head --node-ip-address=172.18.167.77 --redis-password=123 --dashboard-host=0.0.0.0
# On the second server
ray start --node-ip-address 172.18.167.72 --address 172.18.167.77:6379 --redis-password=123
```

Then, on the first server, I ran the following command to start tuning:

```bash
ip_head=172.18.167.77:6379 port=6379 redis_password=123 python mnist.py 
```
A few seconds later, a RuntimeError was raised:

```
Result for TrainMNIST_f8f19_00003:
  date: 2021-05-24_15-59-41
  done: true
  experiment_id: d6b12886023e4faba557a5eb683f9724
  hostname: amax
  iterations_since_restore: 4
  mean_accuracy: 0.0875
  node_ip: 172.18.167.77
  pid: 15999
  time_since_restore: 1.495924949645996
  time_this_iter_s: 0.2805964946746826
  time_total_s: 1.495924949645996
  timestamp: 1621843181
  timesteps_since_restore: 0
  training_iteration: 4
  trial_id: f8f19_00003
  
2021-05-24 15:59:41,683 ERROR trial_runner.py:732 -- Trial TrainMNIST_f8f19_00003: Error processing event.
Traceback (most recent call last):
  File "/home/lyh/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/trial_runner.py", line 716, in _process_trial
    raise RuntimeError(
RuntimeError: Trial TrainMNIST_f8f19_00003 has a non-training future scheduled but 2 results left to process. This should never happen - please file an issue at https://github.com/ray-project/ray/issues
Result for TrainMNIST_f8f19_00003:
  date: 2021-05-24_15-59-41
  done: true
  experiment_id: d6b12886023e4faba557a5eb683f9724
  experiment_tag: 3_lr=0.35202,momentum=0.6942
  hostname: amax
  iterations_since_restore: 4
  mean_accuracy: 0.0875
  node_ip: 172.18.167.77
  pid: 15999
  time_since_restore: 1.495924949645996
  time_this_iter_s: 0.2805964946746826
  time_total_s: 1.495924949645996
  timestamp: 1621843181
  timesteps_since_restore: 0
  training_iteration: 4
  trial_id: f8f19_00003
  
Traceback (most recent call last):
  File "mnist.py", line 58, in <module>
    analysis = tune.run(TrainMNIST,
  File "/home/lyh/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/tune.py", line 520, in run
    runner.step()
  File "/home/lyh/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/trial_runner.py", line 525, in step
    self._process_events(timeout=timeout)  # blocking
  File "/home/lyh/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/trial_runner.py", line 683, in _process_events
    self._execute_action(trial, final_decision)
  File "/home/lyh/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/trial_runner.py", line 974, in _execute_action
    self.trial_executor.continue_training(trial)
  File "/home/lyh/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/ray_trial_executor.py", line 558, in continue_training
    self._train(trial)
  File "/home/lyh/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/ray_trial_executor.py", line 380, in _train
    assert trial.status == Trial.RUNNING, trial.status
AssertionError: ERROR
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

--- a/doc/source/tune/user-guide.rst
+++ b/doc/source/tune/user-guide.rst
@@ -831,7 +831,8 @@ These are the environment variables Ray Tune currently considers:
   is not set, ``~/ray_results`` will be used.
 * **TUNE_RESULT_BUFFER_LENGTH**: Ray Tune can buffer results from trainables before they are passed
   to the driver. Enabling this might delay scheduling decisions, as trainables are speculatively
-  continued. Setting this to ``0`` disables result buffering. Defaults to 1000 (results).
+  continued. Setting this to ``0`` disables result buffering. Defaults to 1000 (results), or to 1 (no buffering)
+  if used with ``checkpoint_at_end``.
 * **TUNE_RESULT_BUFFER_MAX_TIME_S**: Similarly, Ray Tune buffers results up to ``number_of_trial/10`` seconds,
   but never longer than this value. Defaults to 100 (seconds).
 * **TUNE_RESULT_BUFFER_MIN_TIME_S**: Additionally, you can specify a minimum time to buffer results. Defaults to 0.

--- a/python/ray/tune/ray_trial_executor.py
+++ b/python/ray/tune/ray_trial_executor.py
@@ -393,7 +393,7 @@ class RayTrialExecutor(TrialExecutor):
             # If buffer length has not been explicitly set, we determine
             # it automatically
             if buffer_length is None:
-                if trial.checkpoint_at_end or trial.checkpoint_freq:
+                if trial.checkpoint_at_end:
                     # If a trial checkpoint can be triggered externally,
                     # it is not safe to buffer results.
                     buffer_length = 1

--- a/python/ray/tune/ray_trial_executor.py
+++ b/python/ray/tune/ray_trial_executor.py
@@ -25,6 +25,7 @@ from ray.tune.utils.trainable import TrainableUtil
 from ray.tune.trial import Trial, Checkpoint, Location, TrialInfo
 from ray.tune.trial_executor import TrialExecutor
 from ray.tune.utils import warn_if_slow
+from ray.util import log_once
 
 logger = logging.getLogger(__name__)
 
@@ -185,8 +186,10 @@ class RayTrialExecutor(TrialExecutor):
         self.pg_recon_interval = float(
             os.environ.get("TUNE_PLACEMENT_GROUP_RECON_INTERVAL", "5"))
 
-        self._buffer_length = result_buffer_length or int(
+        self._default_buffer_length = result_buffer_length or int(
             os.getenv("TUNE_RESULT_BUFFER_LENGTH", 1000))
+        self._buffer_length = result_buffer_length
+
         self._buffer_min_time_s = float(
             os.getenv("TUNE_RESULT_BUFFER_MIN_TIME_S", 0.))
         self._buffer_max_time_s = float(
@@ -385,8 +388,28 @@ class RayTrialExecutor(TrialExecutor):
             min(self._buffer_max_time_s,
                 len(self._running) // 10))
         with self._change_working_directory(trial):
-            if self._buffer_length > 1:
-                buffer_length = self._buffer_length
+            buffer_length = self._buffer_length
+
+            # If buffer length has not been explicitly set, we determine
+            # it automatically
+            if buffer_length is None:
+                if trial.checkpoint_at_end or trial.checkpoint_freq:
+                    # If a trial checkpoint can be triggered externally,
+                    # it is not safe to buffer results.
+                    buffer_length = 1
+                else:
+                    # Else, use the default buffer length
+                    buffer_length = self._default_buffer_length
+            else:
+                if trial.checkpoint_at_end:
+                    if log_once("trial_executor_buffer_checkpoint"):
+                        logger.warning(
+                            "You passed `checkpoint_at_end` to `tune.run()`, "
+                            "but still requested buffered training. "
+                            "If used with a custom stopper or early stopping, "
+                            "checkpoints may be created later than desired.")
+
+            if buffer_length > 1:
                 if trial.checkpoint_freq > 0:
                     buffer_length = min(buffer_length, trial.checkpoint_freq)
                 remote = trial.runner.train_buffered.remote(

--- a/python/ray/tune/tests/test_trial_runner_3.py
+++ b/python/ray/tune/tests/test_trial_runner_3.py
@@ -12,6 +12,7 @@ import ray
 from ray.rllib import _register_all
 
 from ray.tune import TuneError
+from ray.tune.ray_trial_executor import RayTrialExecutor
 from ray.tune.result import TRAINING_ITERATION
 from ray.tune.schedulers import TrialScheduler, FIFOScheduler
 from ray.tune.experiment import Experiment
@@ -635,6 +636,72 @@ class TrialRunnerTest3(unittest.TestCase):
         runner.step()  # process save
         self.assertEqual(trial.last_result[TRAINING_ITERATION], 9)
         self.assertEqual(num_checkpoints(trial), 3)
+
+    def testCheckpointAtEndNotBuffered(self):
+        os.environ["TUNE_RESULT_BUFFER_LENGTH"] = "7"
+        os.environ["TUNE_RESULT_BUFFER_MIN_TIME_S"] = "0.5"
+
+        def num_checkpoints(trial):
+            return sum(
+                item.startswith("checkpoint_")
+                for item in os.listdir(trial.logdir))
+
+        ray.init(num_cpus=2)
+
+        trial = Trial(
+            "__fake",
+            checkpoint_at_end=True,
+            stopping_criterion={"training_iteration": 4})
+        runner = TrialRunner(
+            local_checkpoint_dir=self.tmpdir, checkpoint_period=0)
+        runner.add_trial(trial)
+
+        runner.step()  # start trial
+
+        runner.step()  # run iteration 1
+        self.assertEqual(trial.last_result[TRAINING_ITERATION], 1)
+        self.assertEqual(num_checkpoints(trial), 0)
+
+        runner.step()  # run iteration 2
+        self.assertEqual(trial.last_result[TRAINING_ITERATION], 2)
+        self.assertEqual(num_checkpoints(trial), 0)
+
+        runner.step()  # run iteration 3
+        self.assertEqual(trial.last_result[TRAINING_ITERATION], 3)
+        self.assertEqual(num_checkpoints(trial), 0)
+
+        runner.step()  # run iteration 4
+        self.assertEqual(trial.last_result[TRAINING_ITERATION], 4)
+        self.assertEqual(num_checkpoints(trial), 1)
+
+    def testCheckpointAtEndBuffered(self):
+        os.environ["TUNE_RESULT_BUFFER_LENGTH"] = "7"
+        os.environ["TUNE_RESULT_BUFFER_MIN_TIME_S"] = "0.5"
+
+        def num_checkpoints(trial):
+            return sum(
+                item.startswith("checkpoint_")
+                for item in os.listdir(trial.logdir))
+
+        ray.init(num_cpus=2)
+
+        trial = Trial(
+            "__fake",
+            checkpoint_at_end=True,
+            stopping_criterion={"training_iteration": 4})
+        # Force result buffer length
+        runner = TrialRunner(
+            local_checkpoint_dir=self.tmpdir,
+            checkpoint_period=0,
+            trial_executor=RayTrialExecutor(result_buffer_length=7))
+        runner.add_trial(trial)
+
+        runner.step()  # start trial
+        self.assertEqual(num_checkpoints(trial), 0)
+
+        runner.step()  # run iterations 1-7
+        self.assertEqual(trial.last_result[TRAINING_ITERATION], 7)
+        self.assertEqual(num_checkpoints(trial), 1)
 
     def testUserCheckpoint(self):
         os.environ["TUNE_RESULT_BUFFER_LENGTH"] = "1"  # Don't finish early

--- a/python/ray/tune/trial_runner.py
+++ b/python/ray/tune/trial_runner.py
@@ -729,12 +729,16 @@ class TrialRunner:
                         # non-training future (e.g. a save) was scheduled.
                         # We do not allow processing more results then.
                         if i < len(results) - 1:
-                            raise RuntimeError(
-                                f"Trial {trial} has a non-training future "
-                                f"scheduled but {len(results)-i} results "
-                                f"left to process. This should never "
-                                f"happen - please file an issue at "
-                                f"https://github.com/ray-project/ray/issues")
+                            if log_once("trial_runner_buffer_checkpoint"):
+                                logger.warning(
+                                    f"Trial {trial} has a non-training future "
+                                    f"scheduled but {len(results)-i} results "
+                                    f"left to process. This means that a "
+                                    f"checkpoint was requested, but buffered "
+                                    f"training was continued before it was "
+                                    f"saved. Consider using non-buffered "
+                                    f"training by setting the env variable "
+                                    f"`TUNE_RESULT_BUFFER_LENGTH=1`.")
                     elif decision == TrialScheduler.STOP:
                         # If the decision is to stop the trial,
                         # ignore all results that came after that.

